io/src/lib.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+#![doc = include_str!("../README.md")]
+#![deny(missing_docs, missing_debug_implementations)]
+
+//! Module for providing stderr IO from an enclave.
+//!
+//! By default stderr from an enclave will be directed to the untrusted (host)
+//! stderr. Consumers can redirect this stream by setting the [`WriteAll`]
+//! function via [`stderr_write_all`].
+
+use once_cell::sync::Lazy;
+use std::ffi::c_void;
+use std::io::Write;
+use std::slice;
+use std::sync::Mutex;
+
+/// A function that that writes the entire contents of the provided buffer.
+///
+/// This is meant be a stand alone version of [`std::io::Writer::write_all`].
+pub type WriteAll = dyn Fn(&[u8]);
+
+/// Specify the function to use for stderr messages
+pub fn stderr_write_all(write_all: &'static WriteAll) {
+    let mut stderr = STDERR.lock().expect("Mutex has been poisoned");
+    stderr.write_all = write_all;
+}
+
+/// Wrap the `write_all` function in a struct so that we can give the `Send`
+/// trait.
+struct Stream {
+    write_all: &'static dyn Fn(&[u8]),
+}
+
+/// SAFETY: The [`Stream`] is local to this crate and will be enclosed in a
+/// Mutex so is safe to make `Send`.
+unsafe impl Send for Stream {}
+
+/// The stderr stream to use for the `ocall_stderr`
+static STDERR: Lazy<Mutex<Stream>> = Lazy::new(|| {
+    Mutex::new(Stream {
+        write_all: &default_stderr_write_all,
+    })
+});
+
+/// A ['WriteAll] function that directs to [`std::io::stderr`]
+fn default_stderr_write_all(stderr: &[u8]) {
+    std::io::stderr()
+        .write_all(stderr.as_ref())
+        .expect("Failed writing to stderr");
+}
```

Comment:
I'll admit this seems a bit hefty and I'm for ways to shrink it down.
Initially I did this to be able to test and verify the contents of the output. After I thought about it longer consumers of this should be able to direct the enclaves stderr as they see fit, whether it be to the untrusted stderr, or to a logger to keep the two stderr streams separate.